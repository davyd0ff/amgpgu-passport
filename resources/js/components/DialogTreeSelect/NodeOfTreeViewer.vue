<style scoped>
.tree-node {
  position: relative;
}

.tree-node .chevron.shown {
  transform: rotate(-90deg);
}

.tree-node .chevron {
  margin: 0 0 0 0;
  transition: transform 0.2s;
  cursor: pointer;
}

.tree-node > label > a > i.material-icons {
  position: absolute;
  right: -10px;
  height: 24px;
  line-height: 1;
  width: 24px;
  color: rgba(0, 0, 0, 0.54);
}
</style>

<template>
  <div v-if="isDisplayed">
    <div v-bind:style="{ ...styles, paddingLeft: `${20}px` }">
      <p class="tree-node">
        <label>
          <input
            type="checkbox"
            class="filled-in"
            v-bind:checked="currentChecked"
            v-on:change="checkboxClicked"
          />
          <span style="color: #212121">{{ node | title }}</span>
          <a
            v-if="nodes.length > 0"
            style="color: #212121"
            v-on:click.stop.prevent="onShowChildren"
          >
            <i
              class="material-icons chevron"
              v-bind:class="{ shown: isShownNodes }"
              >chevron_left</i
            >
          </a>
        </label>
      </p>
      <!-- todo develop: трабла следующая: 2. Если ФЛ будет в нескольких группах, то как его уведомлять? -->
      <!-- допустим в ГруппаА и в ГруппаБ будет СтудентМ -->
      <!-- Если выделить ГруппуА, а затем выделить ГруппуБ и снять выделение с ГруппыА-->
      <!-- то если массив выделенных ФЛ будт distinct то, СтудентМ не попадет в набор -->
      <!-- решение - делать distinct перед отправкой на сервер -->
      <p class="tree-nodes" v-bind:class="`children-select-${level}`">
        <node-of-tree-viewer
          v-for="(node, index) in nodes"
          v-bind:key="`${level}${index}`"
          v-bind:level="level + 1"
          v-bind:node="node"
          v-bind:is-displayed="isShownNodes"
          v-bind:parent-checked="currentChecked"
          v-bind:depth-displayed="depthDisplayed"
          v-bind:styles="styles"
          v-bind:toggle="toggle"
        />
      </p>
    </div>
  </div>
</template>

<script>
export default {
  name: 'NodeOfTreeViewer',
  props: {
    level: { type: Number, default: 0 },
    depthDisplayed: { type: Number, default: 0 },
    isDisplayed: { type: Boolean, default: false },
    parentChecked: { type: Boolean, default: false },
    node: { type: Object, default: null },
    styles: { type: Object, default: () => ({}) },
    toggle: { type: Function, default: (m, a) => {} },
  },
  data() {
    return {
      currentChecked: this.parentChecked,
      isShownNodes: this.level + 1 <= this.depthDisplayed,
    };
  },
  computed: {
    nodes() {
      return this.node?.nodes ?? [];
    },
  },
  watch: {
    parentChecked() {
      this.currentChecked = this.parentChecked;
    },
    currentChecked(value) {
      if (this.node && this.node.member) {
        this.toggle(this.node.member, value);
      }
    },
  },
  filters: {
    title(node) {
      if (node.name) {
        return node.name;
      }
      if (node.member && node.member.name) {
        return node.member.name;
      }

      return 'Empty';
    },
  },
  methods: {
    checkboxClicked() {
      this.currentChecked = !this.currentChecked;
    },
    onShowChildren() {
      this.isShownNodes = !this.isShownNodes;
    },
  },
};
</script>
