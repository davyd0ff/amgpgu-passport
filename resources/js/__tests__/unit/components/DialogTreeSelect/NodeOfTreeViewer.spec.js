import { mount } from '@vue/test-utils';
import NodeOfTreeViewer from '@/components/DialogTreeSelect/NodeOfTreeViewer.vue';

describe('NodeOfTreeViewer.vue', () => {
  const node = {
    name: '1',
    nodes: [
      {
        name: '2-1',
        nodes: [{ name: '3-1' }],
      },
      { name: '2-2' },
    ],
  };

  it('render with depth equals 0', () => {
    const depthDisplayed = 0;
    const isDisplayed = true;

    const wrapper = mount(NodeOfTreeViewer, {
      propsData: {
        depthDisplayed,
        isDisplayed,
        node,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('render with depth equals 1', () => {
    const depthDisplayed = 1;
    const isDisplayed = true;

    const wrapper = mount(NodeOfTreeViewer, {
      propsData: {
        depthDisplayed,
        isDisplayed,
        node,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
