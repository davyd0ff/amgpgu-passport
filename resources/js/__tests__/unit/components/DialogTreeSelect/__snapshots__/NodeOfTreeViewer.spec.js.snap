// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeOfTreeViewer.vue render with depth equals 0 1`] = `
<div>
  <div
    style="padding-left: 20px;"
  >
    <p
      class="tree-node"
    >
      <label>
        <input
          class="filled-in"
          type="checkbox"
        />
         
        <span
          style="color: rgb(33, 33, 33);"
        >
          1
        </span>
         
        <a
          style="color: rgb(33, 33, 33);"
        >
          <i
            class="material-icons chevron"
          >
            chevron_left
          </i>
        </a>
      </label>
    </p>
     
    <p
      class="tree-nodes children-select-0"
    >
      <!---->
      <!---->
    </p>
  </div>
</div>
`;

exports[`NodeOfTreeViewer.vue render with depth equals 1 1`] = `
<div>
  <div
    style="padding-left: 20px;"
  >
    <p
      class="tree-node"
    >
      <label>
        <input
          class="filled-in"
          type="checkbox"
        />
         
        <span
          style="color: rgb(33, 33, 33);"
        >
          1
        </span>
         
        <a
          style="color: rgb(33, 33, 33);"
        >
          <i
            class="material-icons chevron shown"
          >
            chevron_left
          </i>
        </a>
      </label>
    </p>
     
    <p
      class="tree-nodes children-select-0"
    >
      <div>
        <div
          style="padding-left: 20px;"
        >
          <p
            class="tree-node"
          >
            <label>
              <input
                class="filled-in"
                type="checkbox"
              />
               
              <span
                style="color: rgb(33, 33, 33);"
              >
                2-1
              </span>
               
              <a
                style="color: rgb(33, 33, 33);"
              >
                <i
                  class="material-icons chevron"
                >
                  chevron_left
                </i>
              </a>
            </label>
          </p>
           
          <p
            class="tree-nodes children-select-1"
          >
            <!---->
          </p>
        </div>
      </div>
      <div>
        <div
          style="padding-left: 20px;"
        >
          <p
            class="tree-node"
          >
            <label>
              <input
                class="filled-in"
                type="checkbox"
              />
               
              <span
                style="color: rgb(33, 33, 33);"
              >
                2-2
              </span>
               
              <!---->
            </label>
          </p>
           
          <p
            class="tree-nodes children-select-1"
          />
        </div>
      </div>
    </p>
  </div>
</div>
`;
